### PR TITLE
Move management of calls of "cmake --build" to setup_helper/cmake.py and refactor as a CMake class

### PR DESCRIPTION
**Pull request by xuhdev**

### CODE REVIEW

File: tools/build_pytorch_libs.py [hunk 1]
```diff
@@ -61,47 +58,16 @@ def build_caffe2(version,
...
-        os.remove(cmake_cache_file)
-    if not os.path.exists(cmake_cache_file) or (
-            USE_NINJA and not os.path.exists(ninja_build_file)):
-        cmake.run(version,
-                  cmake_python_library,
-                  build_python,
-                  build_test,
-                  build_dir,
-                  my_env)
+    c = CMake(build_dir)
```
Comment (peterjc123) [inline tools/build_pytorch_libs.py]: nit: use a better name here

File: tools/setup_helpers/cmake.py [hunk 2]
```diff
@@ -50,210 +52,252 @@ def _mkdir_p(d):
              _which('ninja') is not None)
 
 
-def get_version(cmd):
-    "Returns cmake version."
+class CMake:
+    "Manage cmake."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: should be `Manages CMake` or `A python wrapper for CMake`.

File: tools/setup_helpers/cmake.py [hunk 3]
```diff
@@ -1,9 +1,11 @@
+"Manage CMake."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: and here

File: tools/setup_helpers/cmake.py [hunk 4]
```diff
@@ -50,210 +52,252 @@ def _mkdir_p(d):
...
+        elif REL_WITH_DEB_INFO:
+            self._build_type = "RelWithDebInfo"
+        else:
+            self._build_type = "Release"
 
-def get_command():
-    "Returns cmake command."
+    @staticmethod
+    def _get_cmake_command():
+        "Return cmake command."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: "Returns"

File: tools/setup_helpers/cmake.py [hunk 5]
```diff
@@ -50,210 +52,252 @@ def _mkdir_p(d):
...
+        # and put the directory as the last element. Otherwise, these flags
+        # may not be passed correctly.
+        # Reference:
+        # 1. https://cmake.org/cmake/help/latest/manual/cmake.1.html#synopsis
+        # 2. https://stackoverflow.com/a/27169347
+        args.append(base_dir)
+        self.run(args, env=my_env)
+
+    def build(self, my_env):
+        "Run cmake to build binaries."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: and here

File: tools/setup_helpers/cmake.py [hunk 6]
```diff
@@ -50,210 +52,252 @@ def _mkdir_p(d):
...
+
+    @staticmethod
+    def defines(args, **kwargs):
+        "Add definitions to a cmake argument list."
+        for key, value in sorted(kwargs.items()):
+            if value is not None:
+                args.append('-D{}={}'.format(key, value))
+
+    def generate(self, version, cmake_python_library, build_python, build_test, my_env, rerun):
+        "Run cmake to generate native build files."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: and here

File: tools/setup_helpers/cmake.py [hunk 7]
```diff
@@ -50,210 +52,252 @@ def _mkdir_p(d):
...
+    def _get_version(cmd):
+        "Returns cmake version."
+
+        for line in check_output([cmd, '--version']).decode('utf-8').split('\n'):
+            if 'version' in line:
+                return LooseVersion(line.strip().split(' ')[2])
+        raise RuntimeError('no version found')
+
+    def run(self, args, env):
+        "Run cmake with arguments and an environment."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: and here

File: tools/setup_helpers/cmake.py [hunk 8]
```diff
@@ -50,210 +52,252 @@ def _mkdir_p(d):
...
+
+        command = [self._cmake_command] + args
+        print(' '.join(command))
+        check_call(command, cwd=self._build_dir, env=env)
+
+    @staticmethod
+    def defines(args, **kwargs):
+        "Add definitions to a cmake argument list."
```
Comment (peterjc123) [inline tools/setup_helpers/cmake.py]: nit: and here

**Review by peterjc123 — COMMENTED** [review]
Sorry for the late review. I didn't get why do we need this refactor? Could you explain this a little bit?

**Comment by ezyang** [comment]
Yeah, it would be really helpful to know the motivation behind the refactor. I personally think that one of our big antipatterns is how long our cmake invocations are; we really should strive to have as few arguments to cmake as possible (and not make it easier to add more and more arguments.)

**Comment by xuhdev** [comment]
The purpose of this PR is a step toward better integration of the Python setup scripts and CMake. Ultimately, I would like to allow build options to be directly specified using options specified in CMakeLists.txt (without explicitly being processed in Python setup scripts), allow users to adjust build options directly via cmake-gui/ccmake/editing CMakeCache.txt (currently users must take care of both options in CMakeCache.txt and environmental variables passed to setup.py and ensure they are consistent upon rebuild), and remove a lot of system and library environment checks in Python setup scripts while CMake can already handle them pretty well.

This PR makes it easier to manage CMake related code in Python setup scripts and paves a path towards these objectives. A well structured code base is important, because the next step is probably to look at one build option at a time PyTorch currently explicitly supports in setup.py.

@ezyang I agree with you that there are many issues with current cmake invocation. The objective I described above subsumes many cmake invocation issues, because only what users have specified will be sent to CMake as arguments.

**Comment by xuhdev** [comment]
@ezyang I should also stress that to address the long CMake invocation argument list, as stated above, I have to look into each build option one at a time. I would focus on those once we have some basic structure in Python setup scripts better organized.

File: tools/build_pytorch_libs.py (hunk 1 above)
Comment (xuhdev) [reply tools/build_pytorch_libs.py]: I fixed all the other comments. Do you think `CMakeManager` is better?

Comment (peterjc123) [reply tools/build_pytorch_libs.py]: No, I mean the name of the variable `c`. Would you please try to make it a bit more meaningful?

Comment (xuhdev) [reply tools/build_pytorch_libs.py]: Indeed, can't believe I gave such a name. Updated.

**Review by peterjc123 — APPROVED** [review]
LGTM, waiting for @ezyang's response

**Review by ezyang — APPROVED** [review]
OK sure.

**Comment by ezyang** [comment]
Things that you don't have to do this PR, but to think about for the future as you keep working on this code (a lot of these are preexisting conditions):

Python nits:

1. More docs: for every function, write a docblock that explains what the expected type and semantics of each argument
2. Let's try to use keyword arguments for long argument lists, like the invocation of `CMake.generate`

General design thoughts:

1. For BC reasons, we do need to keep supporting passing environment variables to invocations of `setup.py` (also, it is the easiest way to make `setup.py` invocations configurable in any case), so it will be worth thinking what the terminal state of the code will be, given this constraint.
2. We should figure out some sort of CI-based testing strategy for direct cmake invocation after the fact. Maybe something as simple as running ninja after setup.py and expecting this to be no-op.

**Comment by xuhdev** [comment]
@ezyang Thanks! Will keep these in mind.

**Pull request closed without merging**